--- a/src/rdkafka_ssl.c	2019-08-29 17:10:11.000000000 -0700
+++ b/src/rdkafka_ssl.c	2019-08-29 17:41:26.000000000 -0700
@@ -1024,7 +1024,7 @@
         int r;
         SSL_CTX *ctx;
 
-#if OPENSSL_VERSION_NUMBER >= 0x10100000
+#if 0
         rd_kafka_dbg(rk, SECURITY, "OPENSSL", "Using OpenSSL version %s "
                      "(0x%lx, librdkafka built with 0x%lx)",
                      OpenSSL_version(OPENSSL_VERSION),
